### PR TITLE
CosmosNullReferenceException: Refactors CosmosNullReferenceException to pass along InnerException property on parent NullReferenceException

### LINKED ISSUE #3705: Refactor CosmosNullReferenceException to capture the Inner Exception in Constructor
- Refactor `CosmosNullReferenceException` to capture the inner exception that is thrown internally, so that the root cause for the null reference exception could be troubleshooted.

### PR DESCRIPTION
**Pull request by jeet1995**
## Description

Sets the `InnerException` property on the base `NullReferenceException` when instantiating `CosmosNullReferenceException`. 

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Closing issues

To automatically close an issue: closes #3705

### CODE REVIEW

**Review by philipthomas-MSFT — DISMISSED** [review]

File: Microsoft.Azure.Cosmos/src/Resource/ClientContextCore.cs [hunk 1]
```diff
@@ -531,6 +531,7 @@ protected virtual void Dispose(bool disposing)
                 catch (NullReferenceException nullRefException) when (!(nullRefException is CosmosNullReferenceException))
                 {
                     CosmosNullReferenceException nullException = new CosmosNullReferenceException(
+                        nullRefException.Message,
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Resource/ClientContextCore.cs:534]: Why is this needed? The nullRefException object is already being passed down, you can extract the message from it in the constructor of the `CosmosNullReferenceException` class.

File: Microsoft.Azure.Cosmos/src/Resource/CosmosExceptions/CosmosNullReferenceException.cs [hunk 2]
```diff
@@ -24,10 +24,12 @@ internal class CosmosNullReferenceException : NullReferenceException
         /// Create an instance of CosmosNullReferenceException
         /// </summary>
         internal CosmosNullReferenceException(
+            String message,
             NullReferenceException originalException,
-            ITrace trace)
+            ITrace trace) 
+            : base(message ?? throw new ArgumentNullException(nameof(message)), originalException ?? throw new ArgumentNullException(nameof(originalException)))
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Resource/CosmosExceptions/CosmosNullReferenceException.cs:30]:
```suggestion
            : base(originalException?.Message ?? throw new ArgumentNullException(nameof(originalException)), originalException ?? throw new ArgumentNullException(nameof(originalException)))
```

File: Microsoft.Azure.Cosmos/src/Resource/CosmosExceptions/CosmosNullReferenceException.cs [hunk 3]
```diff
@@ -24,10 +24,12 @@ internal class CosmosNullReferenceException : NullReferenceException
         /// Create an instance of CosmosNullReferenceException
         /// </summary>
         internal CosmosNullReferenceException(
+            String message,
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Resource/CosmosExceptions/CosmosNullReferenceException.cs:27]:
```suggestion
```

Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Resource/CosmosExceptions/CosmosNullReferenceException.cs:27]: You can use `originalException.Message` to obtain the message without any new extra parameters

**Review by ealsur — CHANGES_REQUESTED** [review]
Thanks for taking care of this, only a few small comments :)

**Review by FabianMeiswinkel — DISMISSED** [review]
LGTM

**Review by philipthomas-MSFT — APPROVED** [review]

File: Microsoft.Azure.Cosmos/src/Resource/CosmosExceptions/CosmosNullReferenceException.cs [hunk 4]
```diff
@@ -25,9 +25,10 @@ internal class CosmosNullReferenceException : NullReferenceException
         /// </summary>
         internal CosmosNullReferenceException(
             NullReferenceException originalException,
-            ITrace trace)
+            ITrace trace) 
+            : base(originalException?.Message ?? throw new ArgumentNullException(nameof(originalException)), originalException ?? throw new ArgumentNullException(nameof(originalException)))
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Resource/CosmosExceptions/CosmosNullReferenceException.cs:29]: Probability of Null passing for exception is very low. And the checks are unnecessary.

If the base type supports null then why not let it flow?

**Review by kirankumarkolli — APPROVED** [review]

File: Microsoft.Azure.Cosmos/src/Resource/CosmosExceptions/CosmosNullReferenceException.cs (hunk 4 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Resource/CosmosExceptions/CosmosNullReferenceException.cs:29]: But in our case, it would be unexpected, right? Practically not possible on the existing code flow but if someone in the future modifies the flow and forgets to pass the originating exception, it would be good to catch it?

Comment (jeet1995) [reply Microsoft.Azure.Cosmos/src/Resource/CosmosExceptions/CosmosNullReferenceException.cs:29]: The base CTOR overloads with `Message` and `InnerException` arguments accept null through the type hierarchy. Throwing in derived class is necessary IMO.
